// @spore-sdk/core/lib/api/composed/spore/destroySpore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.destroySpore = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const helpers_1 = require("../../../helpers");
const spore_1 = require("../../joints/spore");
async function destroySpore(props) {
    // Env
    const config = props.config;
    const indexer = new lumos_1.Indexer(config.ckbIndexerUrl, config.ckbNodeUrl);
    // Get TransactionSkeleton
    let txSkeleton = lumos_1.helpers.TransactionSkeleton({
        cellProvider: indexer,
    });
    // Inject live spore to Transaction.inputs
    const sporeCell = await (0, spore_1.getSporeCellByOutPoint)(props.sporeOutPoint, config);
    const injectLiveSporeCellResult = await (0, spore_1.injectLiveSporeCell)({
        sporeCell,
        txSkeleton,
        config,
    });
    txSkeleton = injectLiveSporeCellResult.txSkeleton;
    // Inject needed capacity and pay fee
    const injectCapacityAndPayFeeResult = await (0, helpers_1.injectCapacityAndPayFee)({
        txSkeleton,
        changeAddress: props.changeAddress,
        fromInfos: props.fromInfos,
        fee: lumos_1.BI.from(0),
        config,
    });
    txSkeleton = injectCapacityAndPayFeeResult.txSkeleton;
    return {
        txSkeleton,
        inputIndex: injectLiveSporeCellResult.inputIndex,
    };
}
exports.destroySpore = destroySpore;
