// @spore-sdk/core/lib/api/composed/cluster/createCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCluster = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const helpers_1 = require("../../../helpers");
const cluster_1 = require("../../joints/cluster");
async function createCluster(props) {
    // Env
    const config = props.config;
    const indexer = new lumos_1.Indexer(config.ckbIndexerUrl, config.ckbNodeUrl);
    // Get TransactionSkeleton
    let txSkeleton = lumos_1.helpers.TransactionSkeleton({
        cellProvider: indexer,
    });
    // Generate and inject cluster cell
    const injectNewClusterResult = (0, cluster_1.injectNewClusterOutput)({
        txSkeleton,
        ...props,
    });
    txSkeleton = injectNewClusterResult.txSkeleton;
    // Inject needed capacity and pay fee
    const injectCapacityAndPayFeeResult = await (0, helpers_1.injectCapacityAndPayFee)({
        txSkeleton,
        changeAddress: props.changeAddress,
        fromInfos: props.fromInfos,
        fee: lumos_1.BI.from(0),
        config,
    });
    txSkeleton = injectCapacityAndPayFeeResult.txSkeleton;
    // Generate and inject cluster ID
    txSkeleton = (0, cluster_1.injectClusterIds)({
        clusterOutputIndices: [injectNewClusterResult.outputIndex],
        txSkeleton,
        config,
    });
    return {
        txSkeleton,
        outputIndex: injectNewClusterResult.outputIndex,
    };
}
exports.createCluster = createCluster;
